cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+         * There is no transition from this state to a new one.
+         */
+        Ready
+    }
+
+    enum RebalanceAnnotation {
+        /**
+         * No annotation set on the rebalance resource.
+         */
+        none,
+        /**
+         * Used to approve a rebalance proposal and trigger the actual rebalancing.
+         * This value should only be used when in the {@code ProposalReady} state.
+         */
+        approve,
+        /**
+         * Used to stop a request for an actual ongoing rebalancing.
+         * This value should only be used when in the {@code Rebalancing} state.
+         */
+        stop,
+        /**
+         * Used to refresh a ready rebalance proposal or to restart a stopped request for getting a rebalance proposal.
+         * This value should only be used when in the {@code ProposalReady} or {@code Stopped} states.
+         */
+        refresh,
+        /**
+         * Any other unsupported/unknown annotation value.
+         */
+        unknown
+    }
+
+    private Future<Void> reconcile(Reconciliation reconciliation, String host, CruiseControlApi apiClient, KafkaRebalance clusterRebalance, State currentState, RebalanceAnnotation rebalanceAnnotation) {
+        RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder = new RebalanceOptions.RebalanceOptionsBuilder();
+        if (clusterRebalance.getSpec().getGoals() != null) {
+            rebalanceOptionsBuilder.withGoals(clusterRebalance.getSpec().getGoals());
+        }
+        if (clusterRebalance.getSpec().isSkipHardGoalCheck()) {
+            rebalanceOptionsBuilder.withSkipHardGoalCheck();
+        }
+        log.info("{}: Rebalance action from state [{}]", reconciliation, currentState);
+
+        return computeNextStatus(reconciliation, host, apiClient, clusterRebalance, currentState, rebalanceAnnotation, rebalanceOptionsBuilder)
+                .compose(desiredStatus -> {
+                    // due to a long rebalancing operation that takes the lock for the entire period, more events related to resource modification could be
+                    // queued with a stale resource (updated by the rebalancing holding the lock), so we need to get the current fresh resource
+                    return kafkaRebalanceOperator.getAsync(reconciliation.namespace(), reconciliation.name())
+                            .compose(freshClusterRebalance -> {
+                                if (freshClusterRebalance != null) {
+                                    return updateStatus(freshClusterRebalance, desiredStatus, kafkaRebalanceOperator, null)
+                                            .compose(updatedClusterRebalance -> {
+                                                log.info("{}: State updated to [{}] with annotation {}={} ",
+                                                        reconciliation,
+                                                        updatedClusterRebalance.getStatus().getConditions().get(0).getType(),
```

Comment:
Similar to the previous comment, this looks like an NPE waiting to happen. However you obtain the condition, you should factor this chain of `get`s into a method.

Reply:
The status was updated at this point with at least/max one condition, so how the NPE could happen? Maybe it's related to the fact that I haven't got yet what you mean in the previous comment. :-)

Reply:
My point is not that the code is _broken_, but that it's _fragile_ to future changes.